### PR TITLE
chore: use stream/sink CAR loader

### PR DESCRIPTION
**Pull request by lemmih**
## Summary of changes

Changes introduced in this pull request:

- Use a sink and a stream when loading CAR files. This makes it easier to control chunking (how much data to send to the DB in one go) and buffering (how many chunks to load in the background while the DB is busy).
- The default chunk size and buffer size give behavior very similar to the old code. Finding better defaults can be done in a separate PR.

## Reference issue to close (if applicable)

Closes

## Other information and links

Benchmarks:

```
Desktop (AMD 3990x, 990 PRO SSD):
    Before:          3:19.22, 5.96 GiB
    After:           2:50.89, 5.50 GiB
```

```
Droplet:
    Before:          8:10.00, 4.88 GiB
    After:           5:57.56, 7.46 GiB
```

```
MacBook Air:
  Before: 187.76s, 5.58 GiB
  After:  205.57s, 6.11 GiB
```

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/genesis/mod.rs [hunk 1]
```diff
@@ -139,42 +142,59 @@ where
...
+                db.put_many_keyed(block.into_iter().map(|block| (block.cid, block.data)))
+            }).await?
+        }
+    };
+
+    car_stream(car_reader)
+        .inspect(|_| n_records += 1)
+        .chunks(1_000)
+        .map(|vec| vec.into_iter().collect::<anyhow::Result<Vec<_>>>())
+        .forward(drain().with(sink).buffer(3))
```
Comment (ruseinov) [inline src/genesis/mod.rs:197]: Curious how you came up with `3` here.

I would argue that both `1_000` and `3` need to become constants with a little doc about their purpose. 

I might be wrong here, but in this base buffer `3` is buffering 3 chunks of `1_000`, correct?

Comment (lemmih) [reply src/genesis/mod.rs:197]: Yes, 3 chunks of `1_000`.

I have very little data supporting these particular values. I did some limited benchmarking and found that 3 buffered chunks of 1_000 key-value pairs worked well most of the time. If the chunk size is too small, ParityDB is unhappy. If the chunk size is too big, we use a lot of memory. If the buffer size is too small, we stall on reading. If the buffer size is too big, we waste memory.

File: src/genesis/mod.rs [hunk 2]
```diff
@@ -139,42 +142,60 @@ where
...
-        tx.send_async((block.cid, block.data)).await?;
-    }
-    drop(tx);
-    write_task.await??;
-    Ok((car_reader.header.roots, n_records))
+
+    let sink = |block: Vec<fvm_ipld_car::Block>| {
+        let db = Arc::clone(&db);
+        async {
+            tokio::task::spawn_blocking(move || {
```
Comment (ruseinov) [inline src/genesis/mod.rs:186]: And this here I assume is done for back-pressure?

File: src/genesis/mod.rs (hunk 1 above)
Comment (ruseinov) [reply src/genesis/mod.rs:202]: Sounds good!

**Review by ruseinov — APPROVED** [review]
This looks good and way more eloquent to me.

**Comment by LesnyRumcajs** [comment]
On my box it's:
```
| branch      | import time |
|-------------|-------------|
| main        | 305s        |
| chunks 1k   | 528s        |
| chunk 40k   | 394s        |
| chunks 200k | 355s        |
```
disk: Kingston KC3000 2TB M.2 Pcie 4.0 NVMe

File: src/genesis/mod.rs (hunk 2 above)
Comment (lemmih) [reply src/genesis/mod.rs:186]: Committing to ParityDB can potentially block indefinitely.

**Comment by aatifsyed** [comment]
Whoops, look like we raced on this: https://github.com/ChainSafe/forest/commit/cd8517e84114b2bc74cad424776898cb544b3ba5#diff-49b3e0dca47375d2ac79eee513ed864b63997ecd184d8101becbe6975e19a3a8

 :D

File: src/genesis/mod.rs [hunk 3]
```diff
@@ -139,42 +142,65 @@ where
...
 }
 
+fn car_stream<R: futures::AsyncRead + Send + Unpin>(
+    mut reader: CarReader<R>,
+) -> impl Stream<Item = anyhow::Result<fvm_ipld_car::Block>> {
+    try_stream! {
+        while let Some(block) = reader.next_block().await? {
+            yield block;
+        }
+    }
```
Comment (aatifsyed) [inline src/genesis/mod.rs:169]: this can be a `futures::stream::unfold` instead of using a macro: 
https://github.com/ChainSafe/forest/blob/fa1126a9596eed07fe0aef951b7225daf4626a8e/benches/car-reader.rs#L29-L30

File: src/genesis/mod.rs [hunk 4]
```diff
@@ -139,42 +142,65 @@ where
...
-    while let Some(block) = car_reader.next_block().await? {
-        debug!("Importing block: {}", block.cid);
-        n_records += 1;
-        tx.send_async((block.cid, block.data)).await?;
-    }
-    drop(tx);
-    write_task.await??;
-    Ok((car_reader.header.roots, n_records))
+
+    let sink = |block: Vec<fvm_ipld_car::Block>| {
```
Comment (aatifsyed) [inline src/genesis/mod.rs:183]: Why introduce a `sink` here? You can just map (concurrently!) into the blockstore:
https://github.com/ChainSafe/forest/blob/cd8517e84114b2bc74cad424776898cb544b3ba5/src/car_stream.rs#L34-L37

Comment (aatifsyed) [reply src/genesis/mod.rs:183]: _Or_ create a struct `CarSink<BlockstoreT>` and `impl Sink for CarSink where BlockstoreT: Blockstore`

File: src/genesis/mod.rs (hunk 3 above)
Comment (lemmih) [reply src/genesis/mod.rs:169]: Nice.

File: src/genesis/mod.rs (hunk 4 above)
Comment (lemmih) [reply src/genesis/mod.rs:183]: Concurrent mapping doesn't work as one might have hoped with non-asynchronous code: It tries to run the code concurrently rather than in parallel, but `put_many_keyed` cannot run concurrently. In this case, `for_each_concurrent` behave exactly as `for_each`, and you don't get the desired buffering behavior.

Refactoring into a `CarSink` sounds like a good idea.

File: src/genesis/mod.rs [hunk 5]
```diff
@@ -186,17 +186,20 @@ where
         },
     );
 
+    let chunk_size = std::env::var("CHUNK_SIZE")?.parse::<usize>()?;
```
Comment (ruseinov) [inline src/genesis/mod.rs:189]: Why not CLI args instead? Would probably be good to have some reasonable defaults too?

File: src/genesis/mod.rs [hunk 6]
```diff
@@ -172,23 +169,22 @@ fn car_stream<R: futures::AsyncRead + Send + Unpin>(
...
-    let db = Arc::new(store);
-
     let car_reader = CarReader::new(reader).await?;
     let roots = car_reader.header.roots.clone();
     let mut n_records = 0;
 
-    let sink = |block: Vec<fvm_ipld_car::Block>| {
-        let db = Arc::clone(&db);
-        async {
+    let sink = futures::sink::unfold(
```
Comment (ruseinov) [inline src/genesis/mod.rs:178]: Nice!

**Comment by lemmih** [comment]
Results from a droplet:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `forest-import-1000-0` | 460.492 ± 1.496 | 458.867 | 461.811 | 1.15 ± 0.05 |
| `forest-import-5000-0` | 463.626 ± 8.641 | 453.677 | 469.252 | 1.16 ± 0.05 |
| `forest-import-10000-0` | 458.160 ± 5.511 | 454.491 | 464.497 | 1.15 ± 0.05 |
| `forest-import-20000-0` | 455.394 ± 8.421 | 450.044 | 465.101 | 1.14 ± 0.05 |
| `forest-import-40000-0` | 439.081 ± 11.196 | 427.520 | 449.872 | 1.10 ± 0.05 |
| `forest-import-200000-0` | 431.252 ± 11.332 | 421.873 | 443.843 | 1.08 ± 0.05 |
| `forest-import-1000-1` | 463.721 ± 4.307 | 458.891 | 467.164 | 1.16 ± 0.05 |
| `forest-import-5000-1` | 454.646 ± 3.110 | 451.883 | 458.014 | 1.14 ± 0.05 |
| `forest-import-10000-1` | 456.409 ± 1.191 | 455.512 | 457.760 | 1.14 ± 0.05 |
| `forest-import-20000-1` | 461.111 ± 4.266 | 456.211 | 463.997 | 1.15 ± 0.05 |
| `forest-import-40000-1` | 440.690 ± 7.126 | 436.432 | 448.917 | 1.10 ± 0.05 |
| `forest-import-200000-1` | 417.070 ± 3.276 | 413.331 | 419.434 | 1.04 ± 0.04 |
| `forest-import-1000-2` | 460.181 ± 6.409 | 454.956 | 467.332 | 1.15 ± 0.05 |
| `forest-import-5000-2` | 460.195 ± 2.087 | 458.789 | 462.592 | 1.15 ± 0.05 |
| `forest-import-10000-2` | 462.527 ± 7.653 | 454.838 | 470.144 | 1.16 ± 0.05 |
| `forest-import-20000-2` | 448.080 ± 5.993 | 441.777 | 453.704 | 1.12 ± 0.05 |
| `forest-import-40000-2` | 437.907 ± 3.067 | 434.417 | 440.175 | 1.10 ± 0.05 |
| `forest-import-200000-2` | 401.267 ± 11.553 | 390.714 | 413.610 | 1.00 ± 0.05 |
| `forest-import-1000-3` | 466.570 ± 6.880 | 458.652 | 471.092 | 1.17 ± 0.05 |
| `forest-import-5000-3` | 457.774 ± 4.235 | 454.594 | 462.581 | 1.15 ± 0.05 |
| `forest-import-10000-3` | 456.932 ± 1.847 | 455.550 | 459.030 | 1.14 ± 0.05 |
| `forest-import-20000-3` | 453.835 ± 3.498 | 450.724 | 457.621 | 1.14 ± 0.05 |
| `forest-import-40000-3` | 439.382 ± 14.580 | 429.766 | 456.158 | 1.10 ± 0.06 |
| `forest-import-200000-3` | 399.766 ± 16.533 | 385.923 | 418.072 | 1.00 |

**Comment by lemmih** [comment]
Results from my desktop:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `forest-import-1000-0` | 165.137 ± 4.740 | 158.060 | 170.747 | 1.00 |
| `forest-import-5000-0` | 172.283 ± 8.144 | 160.141 | 181.027 | 1.04 ± 0.06 |
| `forest-import-10000-0` | 184.627 ± 3.196 | 181.396 | 188.977 | 1.12 ± 0.04 |
| `forest-import-20000-0` | 188.645 ± 4.732 | 180.837 | 192.818 | 1.14 ± 0.04 |
| `forest-import-40000-0` | 193.183 ± 4.745 | 187.587 | 199.244 | 1.17 ± 0.04 |
| `forest-import-200000-0` | 195.772 ± 3.751 | 190.877 | 201.226 | 1.19 ± 0.04 |
| `forest-import-500000-0` | 200.474 ± 4.975 | 196.201 | 208.606 | 1.21 ± 0.05 |
| `forest-import-1000-1` | 198.224 ± 3.736 | 194.700 | 203.822 | 1.20 ± 0.04 |
| `forest-import-5000-1` | 203.418 ± 2.866 | 199.296 | 207.143 | 1.23 ± 0.04 |
| `forest-import-10000-1` | 203.120 ± 4.088 | 198.696 | 209.128 | 1.23 ± 0.04 |
| `forest-import-20000-1` | 201.000 ± 10.880 | 190.569 | 216.758 | 1.22 ± 0.07 |
| `forest-import-40000-1` | 198.633 ± 4.506 | 192.207 | 202.955 | 1.20 ± 0.04 |
| `forest-import-200000-1` | 201.909 ± 7.629 | 194.694 | 212.742 | 1.22 ± 0.06 |
| `forest-import-500000-1` | 201.529 ± 7.101 | 190.545 | 210.444 | 1.22 ± 0.06 |
| `forest-import-1000-2` | 203.184 ± 6.665 | 197.543 | 212.602 | 1.23 ± 0.05 |
| `forest-import-5000-2` | 202.648 ± 10.374 | 189.689 | 216.903 | 1.23 ± 0.07 |
| `forest-import-10000-2` | 204.753 ± 9.200 | 194.155 | 217.358 | 1.24 ± 0.07 |
| `forest-import-20000-2` | 207.397 ± 5.461 | 201.326 | 214.849 | 1.26 ± 0.05 |
| `forest-import-40000-2` | 208.148 ± 7.788 | 197.018 | 218.656 | 1.26 ± 0.06 |
| `forest-import-200000-2` | 207.559 ± 11.221 | 194.097 | 219.429 | 1.26 ± 0.08 |
| `forest-import-500000-2` | 206.661 ± 11.824 | 192.363 | 220.063 | 1.25 ± 0.08 |
| `forest-import-1000-3` | 208.073 ± 4.425 | 203.274 | 215.078 | 1.26 ± 0.05 |
| `forest-import-5000-3` | 204.974 ± 4.552 | 200.739 | 211.984 | 1.24 ± 0.05 |
| `forest-import-10000-3` | 206.022 ± 9.136 | 195.467 | 217.417 | 1.25 ± 0.07 |
| `forest-import-20000-3` | 211.544 ± 9.629 | 201.879 | 227.136 | 1.28 ± 0.07 |
| `forest-import-40000-3` | 203.360 ± 14.617 | 188.599 | 223.000 | 1.23 ± 0.10 |
| `forest-import-200000-3` | 203.711 ± 8.166 | 192.796 | 214.598 | 1.23 ± 0.06 |
| `forest-import-500000-3` | 205.073 ± 6.507 | 198.896 | 214.777 | 1.24 ± 0.05 |


```
Summary
  'forest-import-1000-0' ran
    1.04 ± 0.06 times faster than 'forest-import-5000-0'
    1.12 ± 0.04 times faster than 'forest-import-10000-0'
    1.14 ± 0.04 times faster than 'forest-import-20000-0'
    1.17 ± 0.04 times faster than 'forest-import-40000-0'
    1.19 ± 0.04 times faster than 'forest-import-200000-0'
    1.20 ± 0.04 times faster than 'forest-import-1000-1'
    1.20 ± 0.04 times faster than 'forest-import-40000-1'
    1.21 ± 0.05 times faster than 'forest-import-500000-0'
    1.22 ± 0.07 times faster than 'forest-import-20000-1'
    1.22 ± 0.06 times faster than 'forest-import-500000-1'
    1.22 ± 0.06 times faster than 'forest-import-200000-1'
    1.23 ± 0.07 times faster than 'forest-import-5000-2'
    1.23 ± 0.04 times faster than 'forest-import-10000-1'
    1.23 ± 0.05 times faster than 'forest-import-1000-2'
    1.23 ± 0.10 times faster than 'forest-import-40000-3'
    1.23 ± 0.04 times faster than 'forest-import-5000-1'
    1.23 ± 0.06 times faster than 'forest-import-200000-3'
    1.24 ± 0.07 times faster than 'forest-import-10000-2'
    1.24 ± 0.05 times faster than 'forest-import-5000-3'
    1.24 ± 0.05 times faster than 'forest-import-500000-3'
    1.25 ± 0.07 times faster than 'forest-import-10000-3'
    1.25 ± 0.08 times faster than 'forest-import-500000-2'
    1.26 ± 0.05 times faster than 'forest-import-20000-2'
    1.26 ± 0.08 times faster than 'forest-import-200000-2'
    1.26 ± 0.05 times faster than 'forest-import-1000-3'
    1.26 ± 0.06 times faster than 'forest-import-40000-2'
    1.28 ± 0.07 times faster than 'forest-import-20000-3'
```

They're fairly nonsensical.

File: scripts/db_params_hyperfine.sh [hunk 7]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+
```
Comment (LesnyRumcajs) [inline scripts/db_params_hyperfine.sh:2]:
```suggestion
set -euo pipefail
```

**Review by LesnyRumcajs — APPROVED** [review]

**Comment by aatifsyed** [comment]
> Results from my desktop

- https://github.com/ChainSafe/forest/pull/3090#issuecomment-1621047078

Criterion graphs would be easier to understand for perf tuning :')

File: src/genesis/mod.rs [hunk 8]
```diff
@@ -138,42 +145,71 @@ where
...
+        },
+    );
+
+    // Stream key-value pairs from the CAR file and commit them in chunks. Try
+    // to maintain a buffer of a few chunks to avoid read-stalling.
+    car_stream(car_reader)
+        .inspect(|_| n_records += 1)
+        .chunks(chunk_size)
+        .map(|vec| vec.into_iter().collect::<anyhow::Result<Vec<_>>>())
+        .forward(sink.buffer(buffer_size))
```
Comment (aatifsyed) [inline src/genesis/mod.rs:211]: I don't understand why we have bufferring _and_ chunking - could you add add a doc-comment?

In my head this code is the following, and I don't understand what the introduction of a `sink` does for us
```rust
stream
.chunks(100)
.map(|vec|vec.into_iter().collect())
.try_for_each_concurrent(|vec_of_blocks| tokio::spawn_blocking(||store.put_many_keyed())
```

Comment (lemmih) [reply src/genesis/mod.rs:211]: I did try to explain it in the comments: It's to avoid read-stalling. Reading has higher throughput than writing, but there's quite a bit of latency. A few chunks should be prepared while data is being written to the DB to prevent stalling and hide this latency.

Using `try_for_each_concurrent` with no limits will exhaust the pool of blocking threads. Using a limit is functionally the same as a buffer: The `FuturesUnordered` becomes the buffer. However, this approach requires `DB` to be `Sync,` and the db handle has to be stored in an `Arc`.

TL;DR; `try_for_each_concurrent` can solve the same problem but I feel a sink is prettier (fewer constraints, same number of LoC).

**Review by ruseinov — APPROVED** [review]
